### DISCUSSION TITLE
Unable to use `env` variable inside of matrix definition

### DISCUSSION THREAD

**Discussion by itsayellow** [community/community#26388] [Actions]
<p>In a larger workflow yml file, the following snippet of code causes an error with my workflow.  I can’t figure out why, though.  Is there a valid reason for this to cause an error?  Since my <code>env</code> setting is at the top level of the workflow, it seems like it should be available for defining a matrix inside of a job inside the workflow.</p>
<p>Can somebody help me fix this, or is it a workflow bug?</p>
<pre><code class="lang-auto">name: Test and possibly Publish

on:
  pull_request:
  push:

env:
  default-python: 3.8

jobs:
  tests:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest]
        python-version: [3.6, 3.7, 3.8, 3.9]
        include:
          - os: windows-latest
            python-version: ${{ env.default-python }}
          - os: macos-latest
            python-version: ${{ env.default-python }}

[...]
</code></pre>
<p>Github gives me the following error for the line above</p>
<pre><code class="lang-auto">Invalid workflow file

The workflow is not valid. 
.github/workflows/on-push.yml (Line: 59, Col: 29): Unrecognized named-value: 'env'. Located at position 1 within expression: env.default-python,
.github/workflows/on-push.yml (Line: 61, Col: 29): Unrecognized named-value: 'env'. Located at position 1 within expression: env.default-python
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<p>The documentation on the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#env-context"><code>env</code> context</a> says this:</p>
<blockquote>
<p>You can only use the <code>env</code> context in the value of the <code>with</code> and <code>name</code> keys, or in a step’s <code>if</code> conditional.</p>
</blockquote>
<p>So I’m afraid accessing <code>env</code> variables in a matrix definition is not supported.</p>

**Comment by itsayellow** [comment]
<p>Ah ok!  Thanks for pointing that out.  Everywhere else I tried to use it, it worked, so I didn’t realize it had such limited applicability.</p>

**Comment by itsayellow** [comment]
<p>It would be nice if it were more generally applicable.  It’s nice to use as a “constant” definition at the top of the file, so I don’t have to go chasing down places all throughout the file to change values (and possibly miss some instances where a value needs to be changed.)</p>
